src/eos-updater-fetch.c

```diff
@@ -295,18 +299,460 @@ content_fetch_old (EosUpdater      *updater,
+prepare_flatpaks_to_deploy (OstreeRepo    *repo,
+                            const gchar   *update_id,
+                            GCancellable  *cancellable,
+                            GError       **error)
+{
+  g_autoptr(GHashTable) flatpak_ref_actions_this_commit_wants = NULL;
+  g_autoptr(GHashTable) flatpak_ref_action_progresses = NULL;
+  g_autoptr(GHashTable) relevant_flatpak_ref_actions = NULL;
+  g_autoptr(GPtrArray) flatpaks_to_deploy = NULL;
+  g_autofree gchar *formatted_flatpak_ref_actions_this_commit_wants = NULL;
+  g_autofree gchar *formatted_relevant_flatpak_ref_actions = NULL;
+  g_autofree gchar *formatted_flatpak_ref_actions_progress = NULL;
+
+  flatpak_ref_actions_this_commit_wants = flatpak_ref_actions_for_commit (repo,
+                                                                          update_id,
+                                                                          cancellable,
+                                                                          error);
+
+  if (!flatpak_ref_actions_this_commit_wants)
+    return FALSE;
+
+  formatted_flatpak_ref_actions_this_commit_wants =
+    eos_updater_util_format_all_flatpak_ref_actions ("All flatpak ref actions that this commit wants to have applied on deployment",
+                                                     flatpak_ref_actions_this_commit_wants);
+  g_message ("%s", formatted_flatpak_ref_actions_this_commit_wants);
+
+  flatpak_ref_action_progresses =
+    eos_updater_util_flatpak_ref_action_application_progress_in_state_path (cancellable,
+                                                                            error);
+
+  if (!flatpak_ref_action_progresses)
+    return FALSE;
+
+  formatted_flatpak_ref_actions_progress = eos_updater_util_format_all_flatpak_ref_actions_progresses (flatpak_ref_action_progresses);
+  g_message ("%s", formatted_flatpak_ref_actions_progress);
+
+  /* Filter the flatpak ref actions for the ones which are actually relevant
+   * to this system and figure out which flatpaks need to be pulled from
+   * there */
+  relevant_flatpak_ref_actions = eos_updater_util_filter_for_new_flatpak_ref_actions (flatpak_ref_actions_this_commit_wants,
+                                                                                      flatpak_ref_action_progresses);
+
+  formatted_relevant_flatpak_ref_actions =
+    eos_updater_util_format_all_flatpak_ref_actions ("Flatpak ref actions that need to be prepared while fetching this commit",
+                                                     relevant_flatpak_ref_actions);
+  g_message ("%s", formatted_relevant_flatpak_ref_actions);
+
+  /* Convert the hash table into a single linear array of flatpaks to pull. */
+  flatpaks_to_deploy = eos_updater_util_flatten_flatpak_ref_actions_table (relevant_flatpak_ref_actions);
+
+  if (!pull_flatpaks (flatpaks_to_deploy, cancellable, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gboolean
+content_fetch (EosUpdater      *updater,
```

Comment:
Note for future branches: it’s a lot easier to review stuff like this if the branch is structured as a series of commits which, separately: refactor the existing code without changing its behaviour (e.g. pull `content_fetch()` out of `content_fetch_task()`), add the new code (e.g. `flatpak.c`), and then add calls to the new code into the old code (e.g. the `prepare_flatpaks_to_deploy()` call below). The third step would ideally be as minimal as possible. Then a reviewer who cares about ensuring the existing code continues to work only has to review the first commit in detail — and that first and second commit can potentially be merged in advance of the third one, since they can easily be guaranteed to not break anything.

(I don’t want this to be changed in this branch now; that’s too much rebase faff.)

Reply:
I just pushed the `commit_fetch()`/`commit_fetch_task()` split to master, so you’re going to get some minor rebase problems here.